README.md
# Logical Partitions Plugin for Knife

[![Build Status](https://travis-ci.org/chef/knife-lpar.svg?branch=master)](https://travis-ci.org/chef/knife-lpar)
[![Gem Version](https://badge.fury.io/rb/knife-lpar.svg)](https://badge.fury.io/rb/knife-lpar)

This is the official Chef plugin for managing POWER architecture logical partitions (LPAR).

This plugin gives knife the ability to initialize, start and destroy LPARs.

For more information about LPARs and terminology, see the following: <http://www.redbooks.ibm.com/abstracts/sg247491.html?Open>

WARNING: This code is currently a research project, experimental and provided without warranty or support!

## Installation

### Build Locally

If you would like to build the gem from source locally, please clone this repository on to your local machine and build the gem locally.

```
$ bundle install
$ bundle exec gem install
```

## Subcommands

This plugin provides the following Knife subcommands. Specific command options can be found by invoking the subcommand with a `--help` flag.

### knife lpar create HMC (options)

Creates a new LPAR based on defaults found below. Creates a 50GB backing device (vscsi) with the same name as the LPAR. This is currently not configurable. Once it completes successfully, the LPAR is booted into SMS mode to allow for any extra manual configuration and OS installation.

#### Required

- `HMC`: The name or IP of the Hardware Management Console (example: `hmc01.myurl.com`)
- `--name`: The friendly name of the LPAR (example: `lpar01`)
- `--vios`: The name of the Virtual I/O Server Partition that provides I/O services for the LPARs on the Virtual Server (example: `vios01`)
- `--virtual-server`: The name of the Physical Virtual Server that all of the Partitions are running on (example: `aixvirt01`)
- `--disk-name`: The name of the virtual ISO that has been previously loaded into your Virtual Server's Virtual Media Library to install (example: `AIX6_Disk01`)

#### Optional

- `--profile`: The name of the profile to be used (if not specified, defaults to `--name`)
- `--min-mem`: Minimum Memory in mb (default 1024)
- `--desired-mem`: Desired Memory in mb (default 4096)
- `--max-mem`: Max Memory in mb (default 16384)
- `--min-procs`: Minimum number of allocated physical processors (default 1)
- `--desired-procs`: Desired number of allocated physical processors (default 2)
- `--max-procs`: Max number of allocated physical processors (default 4)
- `--min-proc_units`: Minimum number of processor units (default 1)
- `--desired-proc_units`: Desired number of processor units (default 2)
- `--max-proc_units`: Max number of processor units (default 4)

### knife lpar delete HMC (options)

Removes a specific LPAR from a virtual server and removes any backing devices, logical volumes, virtual optical devices, networking devices, device VIOS mappings, and finally the LPAR itself.

This requires the LPAR to be powered off.

#### Required

- `HMC`: The name or IP of the Hardware Management Console (example: `hmc01.myurl.com`)
- `--name`: The friendly name of the LPAR (example: `lpar01`)
- `--vios`: The name of the Virtual I/O Server Partition that provides I/O services for the LPARs on the Virtual Server (example: `vios01`)
- `--virtual-server`: The name of the Physical Virtual Server that all of the Partitions are running on (example: `aixvirt01`)

### knife lpar list HMC (options)

Lists all LPARs known on a virtual server.

#### Required

- `HMC`: The name or IP of the Hardware Management Console (example: `hmc01.myurl.com`)
- `--virtual-server`: The name of the Physical Virtual Server that all of the Partitions are running on (example: `aixvirt01`)

## Contributing

For information on contributing to this project see <https://github.com/chef/chef/blob/master/CONTRIBUTING.md>

## License and Authors

- Author:: Scott Hain ([shain@chef.io](mailto:shain@chef.io))

```text
Copyright 2014-2016 Chef Software, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# knife-lpar Changelog

## [v0.0.3](https://github.com/chef/mixlib-versioning/tree/v0.0.3) (2018-07-30)

- Removed the runtime dependency on Chef ~11.
- Adding testing in Travis-CI
- Removed outdated contributing docs
- Moved test deps to the Gemfile
- Improved readme content
- Added a Rakefile for testing
- Resolved Chefstyle warnings


lib/chef/knife/lpar_create.rb
#
# Copyright:: Copyright (c) 2014-2016 Chef Software Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require "chef/knife"
require "chef/knife/lpar_base"

class Chef
  class Knife
    class LparCreate < Knife
      include Chef::Knife::LparBase

      banner "knife lpar create HMC [options]"

      option :name,
        :short => "-n NAME",
        :long => "--name",
        :description => "LPAR Name"

      option :profile,
        :short => "-p PROFILE",
        :long => "--profile",
        :description => "LPAR Profile Name"

      option :virtual_server,
        :short => "-v SERVER",
        :long => "--virtual-server",
        :description => "Virtual Server Name"

      option :vios,
        :long => "--vios NAME",
        :description => "Virtual I/O Server LPAR Name"

      option :min_mem,
        :long => "--min-mem MEM",
        :description => "Minimum Memory in mb (default 1024)",
        :default => 1024

      option :desired_mem,
        :long => "--desired-mem MEM",
        :description => "Desired Memory in mb (default 4096)",
        :default => 4096

      option :max_mem,
        :long => "--max-mem MEM",
        :description => "Max Memory in mb (default 16384)",
        :default => 16384

      option :min_procs,
        :long => "--min-procs PROCS",
        :description => "Minimum number of Processors (default 1)",
        :default => 1

      option :desired_procs,
        :long => "--desired-procs PROCS",
        :description => "Desired number of Processors (default 2)",
        :default => 2

      option :max_procs,
        :long => "--max-procs PROCS",
        :description => "Max number of Processors (default 4)",
        :default => 4

      option :min_proc_units,
        :long => "--min-proc_units UNITS",
        :description => "Minimum number of processor units (default 1)",
        :default => 1

      option :desired_proc_units,
        :long => "--desired-proc_units UNITS",
        :description => "Desired number of processor units (default 2)",
        :default => 2

      option :max_proc_units,
        :long => "--max-proc_units UNITS",
        :description => "Max number of processor units (default 4)",
        :default => 4

      option :help,
        :long => "--help",
        :description => "Prints this menu"

      option :disk_name,
        :long => "--disk-name DISK",
        :description => "Disk image name (e.g. AIX_6_1_vol1)"

      #
      # Run the plugin
      #
      def run
        read_and_validate_params
        @password = get_password
        create_lpar
      end

      #
      # Reads the input parameters and validates them.
      # Will exit if it encounters an error
      #
      def read_and_validate_params
        if @name_args.length < 1
          show_usage
          exit 1
        end

        if config[:name].nil? ||
            config[:vios].nil? ||
            config[:virtual_server].nil? ||
            config[:disk_name].nil?
          show_usage
          exit 1
        end

        if config[:profile].nil?
          config[:profile] = config[:name]
        end
      end

      def create_lpar
        Net::SSH.start(@name_args[0], "hscroot", :password => @password) do |ssh|
          # some background checks
          # check for existing lpar with name
          ui.info "Searching for existing lpar with name: #{config[:name]}"
          command = "lssyscfg -m #{config[:virtual_server]} -F name -r lpar | grep #{config[:name]}"
          output = run_remote_command(ssh, command)
          unless output.nil?
            ui.fatal "An lpar already exists with the name #{config[:name]}"
            exit 1
          end
          ui.info "lpar not found, creation imminent"

          # find the last vscsi device number
          ui.info "Looking for next device number in sequence"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"lsdev -type disk -virtual -field name\" | tail -1"
          last_vscsi = run_remote_command(ssh, command)
          ui.info "Found existing device - #{last_vscsi}"

          # use the vscsi number to find the actual physical ID so we can find which vios slot it's in
          ui.info "Finding vios mapping for device - #{last_vscsi}"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"lsdev -dev #{last_vscsi} -field physloc -fmt \\\":\\\"\""
          last_vscsi_phy_loc = run_remote_command(ssh, command)
          prev_loc = last_vscsi_phy_loc.match('.*-C(\d+)-.*')[1]
          ui.info "Found vios mapping #{prev_loc}"
          new_virt_loc = prev_loc.to_i + 1
          ui.info "Will use new mapping #{new_virt_loc}"

          # create the new lpar
          ui.info "Creating new lpar #{config[:name]}"
          command = "mksyscfg -m #{config[:virtual_server]} -r lpar \
-i \"name=#{config[:name]}, \
profile_name=#{config[:profile]}, \
lpar_env=aixlinux, \
min_mem=#{config[:min_mem]}, \
desired_mem=#{config[:desired_mem]}, \
max_mem=#{config[:max_mem]}, \
proc_mode=shared, \
min_procs=#{config[:min_procs]}, \
desired_procs=#{config[:desired_procs]}, \
max_procs=#{config[:max_procs]}, \
min_proc_units=#{config[:min_proc_units]}, \
desired_proc_units=#{config[:desired_proc_units]}, \
max_proc_units=#{config[:max_proc_units]}, \
sharing_mode=uncap, uncap_weight=128, \
boot_mode=norm, max_virtual_slots=10, \
\\\"virtual_eth_adapters=3/0/1//0/0\\\", \
\\\"virtual_scsi_adapters=2/client//#{config[:vios]}/#{new_virt_loc}/1\\\"\""
          output = run_remote_command(ssh, command)
          ui.info "Creation Successful"

          # now we have to figure out what LPAR we just created
          ui.info "Finding vhost name"
          command = "lssyscfg -m #{config[:virtual_server]} --filter \"lpar_names=#{config[:name]}\" -F lpar_id -r lpar"
          output = run_remote_command(ssh, command)
          # and of course it doesn't match, 0 based vs 1 based counting
          vhost = output.to_i - 1
          vhost_name = "vhost#{vhost}"
          ui.info "#{config[:name]} is #{vhost_name}"

          # Add the virtual io server vscsi mapping
          ui.info "Mapping #{new_virt_loc} between #{config[:vios]} and #{config[:name]}"
          command = "chhwres -r virtualio -m #{config[:virtual_server]} -o a -p #{config[:vios]} --rsubtype scsi -s #{new_virt_loc} -a \"adapter_type=server, remote_lpar_name=#{config[:name]}, remote_slot_num=2\""
          output = run_remote_command(ssh, command)
          ui.info "Mapping Successful"

          # make a file backed optical drive
          ui.info "Creating virtual file backed optical device"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"mkvdev -fbo -vadapter #{vhost_name}\""
          vopt_name = run_remote_command(ssh, command).split(" ")[0]
          ui.info "Created device #{vopt_name}"

          # load the iso
          ui.info "Loading disk in optical drive"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"loadopt -vtd #{vopt_name} -disk #{config[:disk_name]}\""
          output = run_remote_command(ssh, command)
          ui.info "Loading Successful"

          # create logical volume
          ui.info "Creating logical volume"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"mklv -lv #{config[:name]} rootvg 50G\""
          lv_name = run_remote_command(ssh, command)
          ui.info "Created logical volume #{lv_name}"

          # attach it
          ui.info "Attaching lv to lpar"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"mkvdev -vdev #{config[:name]} -vadapter #{vhost_name}\""
          vtscsi_name = run_remote_command(ssh, command).split(" ")[0]
          ui.info "Attach Successful as #{vtscsi_name}"

          # save the virtual io server profile
          ui.info "Activating virtual io server profile"
          command = "mksyscfg -r prof -m #{config[:virtual_server]} -o save -p #{config[:vios]} -n `lssyscfg -r lpar -m #{config[:virtual_server]} --filter \"lpar_names=#{config[:vios]}\" -F curr_profile` --force"
          output = run_remote_command(ssh, command)
          ui.info "Activation Successful"

          # reload the lpar so it knows it has new devices
          ui.info "Reload virtual io server to re-read devices"
          command = "viosvrcmd -m #{config[:virtual_server]} -p #{config[:vios]} -c \"cfgdev\""
          output = run_remote_command(ssh, command)
          ui.info "Reload Successful"

          # could start it up here, we'll see
          ui.info "Boot lpar in SMS mode"
          command = "chsysstate -r lpar -m #{config[:virtual_server]} -o on -f #{config[:profile]} -b sms -n #{config[:name]}"
          output = run_remote_command(ssh, command)
          unless output.nil?
            ui.info output.to_s
          end
          ui.info "Boot Successful"
        end
      end

    end
  end
end
